### PR TITLE
bug/EN-3146-Request header by hash when fork is detected

### PR DESCRIPTION
**Pull request by SebastianMarian**

### CODE REVIEW

File: process/sync/basicForkDetector_test.go [hunk 1]
```diff
@@ -217,7 +217,7 @@ func TestBasicForkDetector_CheckForkOnlyOneHeaderOnANonceShouldReturnFalse(t *te
 	bfd, _ := sync.NewBasicForkDetector(rounderMock)
 	_ = bfd.AddHeader(&block.Header{Nonce: 0, PubKeysBitmap: []byte("X")}, []byte("hash1"), process.BHProcessed)
 	_ = bfd.AddHeader(&block.Header{Nonce: 1, PubKeysBitmap: []byte("X")}, []byte("hash2"), process.BHProcessed)
-	forkDetected, lowestForkNonce := bfd.CheckFork()
+	forkDetected, lowestForkNonce, _ := bfd.CheckFork()
 	assert.False(t, forkDetected)
 	assert.Equal(t, uint64(math.MaxUint64), lowestForkNonce)
```
Comment (sasurobert) [inline process/sync/basicForkDetector_test.go]: test added hash.

File: process/sync/shardblock_test.go [hunk 2]
```diff
@@ -930,8 +930,8 @@ func TestBootstrap_SyncBlockShouldCallForkChoice(t *testing.T) {
 	hasher := &mock.HasherMock{}
 	marshalizer := &mock.MarshalizerMock{}
 	forkDetector := &mock.ForkDetectorMock{}
-	forkDetector.CheckForkCalled = func() (bool, uint64) {
-		return true, math.MaxUint64
+	forkDetector.CheckForkCalled = func() (bool, uint64, []byte) {
+		return true, math.MaxUint64, nil
```
Comment (sasurobert) [inline process/sync/shardblock_test.go]: add test for this special scenario as discussed.

File: process/sync/basicForkDetector_test.go (hunk 1 above)
Comment (SebastianMarian) [reply process/sync/basicForkDetector_test.go]: Done

File: process/sync/shardblock_test.go (hunk 2 above)
Comment (SebastianMarian) [reply process/sync/shardblock_test.go]: Done

File: process/sync/shardblock_test.go [hunk 3]
```diff
@@ -1676,15 +1678,163 @@ func TestBootstrap_ShouldReturnTrueWhenNodeIsNotSynced(t *testing.T) {
...
+	hdr2 := block.Header{Nonce: 1, Round: 1, PubKeysBitmap: []byte("B")}
+	hash2 := []byte("hash2")
+
+	blkc := mock.BlockChainMock{}
+	blkc.GetCurrentBlockHeaderCalled = func() data.HeaderHandler {
+		return &hdr2
+	}
+
+	pools := createMockPools()
+	pools.HeadersCalled = func() storage.Cacher {
```
Comment (AdoAdoAdo) [inline process/sync/shardblock_test.go]: I see you are using this function in two tests, please extract it in a separate function and use it in both tests, to remove code duplication

File: process/interface.go [hunk 4]
```diff
@@ -149,7 +149,7 @@ type Bootstrapper interface {
 type ForkDetector interface {
 	AddHeader(header data.HeaderHandler, hash []byte, state BlockHeaderState) error
 	RemoveHeaders(nonce uint64, hash []byte)
-	CheckFork() (bool, uint64)
+	CheckFork() (bool, uint64, []byte)
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/interface.go]: Done

File: process/sync/basicForkDetector.go [hunk 5]
```diff
@@ -294,6 +297,7 @@ func (bfd *basicForkDetector) CheckFork() (bool, uint64) {
 
 		selfHdrInfo = nil
 		lowestRoundInForkNonce = math.MaxUint32
+		hashOfForkNonceWithLowestRound = nil
```
[the remark this answers is not in the document]
Comment (SebastianMarian) [reply process/sync/basicForkDetector.go]: Renamed

File: process/sync/shardblock_test.go (hunk 3 above)
Comment (SebastianMarian) [reply process/sync/shardblock_test.go]: Refactored

**Pull request merged**